### PR TITLE
Codegen: Ignore missing required keys in help mode

### PR DESCRIPTION
**Pull request by kodebach**
If the user starts an application that has required keys with `-h`/`--help`, it would result in an error and the help message wouldn't be printed. This changes that.

It also adds the regenerated man page that is missing from #2942 (should've been part of 
b18447e).

## Basics

- [x] Short descriptions should be in the release notes (added as entry in
      `doc/news/_preparation_next_release.md` which contains `_(my name)_`)
      **Please always add something to the the release notes.**
- [x] Longer descriptions should be in documentation or in design decisions.
- [x] Describe details of how you changed the code in commit messages
      (first line should have `module: short statement` syntax)
- [x] References to issues, e.g. `close #X`, should be in the commit messages.

## Checklist

- [x] I added unit tests for my code
- [x] I fixed all affected documentation
- [x] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [x] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [x] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

- [ ] Documentation is introductory, concise and good to read
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

### CODE REVIEW

**Comment by kodebach** [comment]
jenkins build libelektra please

File: doc/news/_preparation_next_release.md [hunk 1]
```diff
@@ -80,6 +80,10 @@ The following section lists news about the [modules](https://www.libelektra.org/
 
 - We now treat relative paths as relative to `KDB_DB_SPEC` instead of the current working directory. _(Klemens Böswirth)_
 
+### Spec
+
+- There is now the config key `missing/log` that allows logging of all missing `require`d keys. _(Klemens Böswirth)_
```
Comment (markus2330) [inline doc/news/_preparation_next_release.md]: Shouldn't we fail if a required key is missing? Why do we need the log as this should already be part of the error message?

Comment (kodebach) [reply doc/news/_preparation_next_release.md]: 1. `spec` can be configured to ignore some or all errors. I don't really know why, but it's always been the case.
2. This is needed for the case where a program using the codegen API is called in "help mode", i.e. `app -h` or `app --help`. In such a case, we want to ignore missing required keys and simply print the help message.
How the `missing/log` array helps detect this case, is documented in the relevant code:
https://github.com/ElektraInitiative/libelektra/blob/ea74e58203edeaf9175d65f091753d286887da45/src/libs/highlevel/elektra.c#L143-L158
There could technically be other problems from the `spec` plugin as well, but since `gopts` runs first and was successful, there can't be any errors that would break the help message.

Comment (markus2330) [reply doc/news/_preparation_next_release.md]: > spec can be configured to ignore some or all errors. I don't really know why, but it's always been the case.

This is one of spec's anti-features. Thank you for documenting it but I would be in favor of removing these anti-features. If validation fails, the plugin should return an error.

> This is needed for the case where a program using the codegen API is called in "help mode"

Ok. But wouldn't it be better to directly ask for a meta-key which says that we are in the help mode? We want the help to be printed no matter what errors happened in `kdbOpen` or `kdbGet`.

Comment (kodebach) [reply doc/news/_preparation_next_release.md]: > We usually only add warnings during kdbGet()

There are many cases, where `kdbGet()` should return an error. The best example would be a broken storage file. The storage plugin should return an error and `kdbGet()` should abort.

For the `spec` plugin, it probably makes sense to have to option of switching between warnings and errors. In `kdb` (and other configuration tools) warnings would be preferred, as a `kdbGet()` is required before a `kdbSet()` could fix the problems. In applications on the other hand errors are the better solution. They are easier to detect, since `kdbGet()` uses a different return code for errors. The handling of actual errors is also easier in the high-level API (see #2952).

> Can you make a decision of what you wrote here?

Will do.

Comment (markus2330) [reply doc/news/_preparation_next_release.md]: > There are many cases, where kdbGet() should return an error.

I agree. I made a different proposal: #2955

> For the spec plugin, it probably makes sense to have to option of switching between warnings and errors.

I hope this will not be needed once we have #2955 implemented

>   Will do.

:sparkling_heart:

**Comment by kodebach** [comment]
Side note: The fallback help message only works with `app -h` or `app --help`, but not with combined short options like `app -fh`. The reason is simply, that without the spec we don't know which option takes arguments and which doesn't (`app -fh` might be the same as `app -f -h` or `app -f h`). I think this limitation is acceptable.

File: doc/decisions/highlevel_help_message.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+## Constraints
+
+- `elektraOpen` should not return a broken `Elektra` instance.
+- The help message can only be printed, if `elektraOpen` returns an `Elektra` instance and no `ElektraError`.
+
+## Assumptions
+
+- We assume that the application in question was correctly installed.
+- We assume `gopts` was mounted.
```
Comment (markus2330) [inline doc/decisions/highlevel_help_message.md]: Is this already default?

File: src/libs/ease/symbols.map [hunk 3]
```diff
@@ -10,7 +10,9 @@ libelektra_0.8 {
 	elektraKeyGetRelativeName;
 	elektraKsFilter;
 	elektraResolveReference;
+};
 
+libelektra_0.9 {
```
Comment (markus2330) [inline src/libs/ease/symbols.map]: Why this change?

File: src/plugins/gopts/gopts.c [hunk 4]
```diff
@@ -43,6 +72,7 @@ int elektraGOptsGet (Plugin * handle ELEKTRA_UNUSED, KeySet * returned, Key * pa
 			ksNew (30, keyNew ("system/elektra/modules/gopts", KEY_VALUE, "gopts plugin waits for your orders", KEY_END),
 			       keyNew ("system/elektra/modules/gopts/exports", KEY_END),
 			       keyNew ("system/elektra/modules/gopts/exports/get", KEY_FUNC, elektraGOptsGet, KEY_END),
+			       keyNew ("system/elektra/modules/gopts/exports/ishelpmode", KEY_FUNC, elektraGOptsIsHelpMode, KEY_END),
```
Comment (markus2330) [inline src/plugins/gopts/gopts.c]: isinhelpmode or simply helpmode?

File: src/plugins/gopts/gopts.c [hunk 5]
```diff
@@ -34,6 +34,35 @@ static void cleanupEnvp (char ** envp);
...
+	int argc = loadArgs (&argv);
+
+	if (argv == NULL)
+	{
+		return -1;
+	}
+
+	for (int i = 0; i < argc; ++i)
+	{
+		if (strcmp (argv[i], "-h") == 0 || strcmp (argv[i], "--help") == 0)
```
Comment (markus2330) [inline src/plugins/gopts/gopts.c]: I do not like that this is hardcoded here. `-H` is very common (and also used in the `kdb` tool). `-h` has man other meanings.

Furthermore this is not properly parsed, e.g. `-vh` would not be recognized.

File: src/libs/ease/symbols.map (hunk 3 above)
Comment (kodebach) [reply src/libs/ease/symbols.map]: It should have been part of #2916. The functions were not part of any 0.8.x release, so they should not be versioned as `libelektra_0.8`.

If we won't have another 0.9.x release and go straight to 1.0.0, they should actually be in `libelektra_1.0`.

**Comment by markus2330** [comment]
Thank you!

File: doc/decisions/highlevel_help_message.md (hunk 2 above)
Comment (kodebach) [reply doc/decisions/highlevel_help_message.md]: `gopts` is not mounted by default no. However, the file generated by the `highlevel` template, include a `kdbEnsure` contract, that mounts it, if it is missing.

File: src/plugins/gopts/gopts.c (hunk 5 above)
Comment (kodebach) [reply src/plugins/gopts/gopts.c]: > I do not like that this is hardcoded here. `-H` is very common (and also used in the `kdb` tool). `-h` has man other meanings.

Using `-h`/`--help` for help mode is already hard-coded in `elektraGetOpts` (there it is properly parsed).

This function is not actually intended to accurately detect help mode (for that one should still check for the presence of `proc/elektra/gopts/help`. It is only meant in case `kdbGet()` failed.

Apart from `kdb` tool, I have never seen `-H` used in favour of `-h`. The `kdb` tool also takes a very unusual approach, since it opens the man-page instead of printing to stdout/stderr.
In fact most tools I know, use only `--help` (if they support long options).

> Furthermore this is not properly parsed, e.g. `-vh` would not be recognized.

See above https://github.com/ElektraInitiative/libelektra/pull/2944#issuecomment-531577466.

File: src/plugins/gopts/gopts.c (hunk 4 above)
Comment (kodebach) [reply src/plugins/gopts/gopts.c]: Since you already merged this PR: Do you still want the change?

File: doc/decisions/highlevel_help_message.md (hunk 2 above)
Comment (markus2330) [reply doc/decisions/highlevel_help_message.md]: Please copy this sentence into the decision.

File: src/libs/ease/symbols.map (hunk 3 above)
Comment (markus2330) [reply src/libs/ease/symbols.map]: We will definitely have a 0.9 release before 1.0.

File: src/plugins/gopts/gopts.c (hunk 4 above)
Comment (markus2330) [reply src/plugins/gopts/gopts.c]: It was only a question. If you like ishelpmode best, we can keep it.

File: src/plugins/gopts/gopts.c (hunk 5 above)
Comment (markus2330) [reply src/plugins/gopts/gopts.c]: > Apart from kdb tool, I have never seen -H used in favour of -h.

Because of how the cmdline options in Elektra are implemented every option needs to have a short option. `-H` was the most natural one as `-h` is already "human readable".

> The kdb tool also takes a very unusual approach, since it opens the man-page instead of printing to stdout/stderr.

It makes a lot of sense to not have a --help output and a separated man page. E.g. git also does it this way.  

> In fact most tools I know, use only --help (if they support long options).

I would prefer if we only hardcode `--help` and drop support for `-h`.

**Comment by markus2330** [comment]
> Side note: The fallback help message only works with app -h or app --help, but not with combined short options like app -fh. The reason is simply, that without the spec we don't know which option takes arguments and which doesn't (app -fh might be the same as app -f -h or app -f h). I think this limitation is acceptable.

Please make sure that such notes are documented in the documentation. Here nobody will read it.

**Comment by kodebach** [comment]
It is documented in the doxygen comment for `elektraGOptsIsHelpMode`, I just added the note here, in case you miss the comment.

**Comment by markus2330** [comment]
Sorry, I do not see it. You mean src/plugins/gopts/gopts.c line 37-44?

**Comment by kodebach** [comment]
> You mean src/plugins/gopts/gopts.c line 37-44?

Yes
